### PR TITLE
Add better endpoint for string configuration

### PR DESCRIPTION
**Pull request by patrickelectric**
It validates legacy formats as well url formats.
Also allows custom legacy formats to be compatible with previous tools.

### CODE REVIEW

File: src/cli.rs [hunk 1]
```diff
@@ -51,35 +55,21 @@ struct Args {
...
-
-    // Add your custom validation logic here if needed
-    Ok(endpoint)
+fn endpoints_parser(endpoint: &str) -> Result<Arc<dyn drivers::Driver>, String> {
+    drivers::endpoints()
+        .iter()
+        .map(|driver| driver.driver_ext.create_endpoint(endpoint))
+        .filter_map(|driver| driver)
+        .next()
+        .ok_or_else(|| format!("Invalid endpoint: {}", endpoint))
```
Comment (joaoantoniocardoso) [inline src/cli.rs:64]:
```suggestion
        .ok_or_else(|| format!("Invalid endpoint: {endpoint}"))
```

File: src/cli.rs [hunk 2]
```diff
@@ -102,96 +92,68 @@ pub fn init() {
...
 }
 
-// Return a clone of current Args struct
-
-#[instrument(level = "debug")]
-pub fn command_line() -> String {
-    format!("{:#?}", MANAGER.clap_matches)
+#[cfg(test)]
+mod tests {
+    // Note this useful idiom: importing names from outer (for mod tests) scope.
```
Comment (joaoantoniocardoso) [inline src/cli.rs:135]:
```suggestion
```

File: src/cli.rs [hunk 3]
```diff
@@ -51,35 +55,21 @@ struct Args {
...
+        .map(|driver| driver.driver_ext.create_endpoint(endpoint))
+        .filter_map(|driver| driver)
+        .next()
+        .ok_or_else(|| format!("Invalid endpoint: {}", endpoint))
 }
 
-#[derive(Debug)]
 struct Manager {
-    clap_matches: Args,
+    pub clap_matches: Args,
```
Comment (joaoantoniocardoso) [inline src/cli.rs:68]:
```suggestion
    clap_matches: Args,
```
this probably doesn't need to be public

File: src/cli.rs [hunk 4]
```diff
@@ -51,35 +55,21 @@ struct Args {
...
 }
 
 lazy_static! {
-    static ref MANAGER: Arc<Manager> = Arc::new(Manager::new());
+    static ref MANAGER: Arc<RwLock<Manager>> = Arc::new(RwLock::new(Manager::new()));
```
Comment (joaoantoniocardoso) [inline src/cli.rs:72]:
```suggestion
    static ref MANAGER: Arc<Manager> = Arc::new(Manager::new());
```

Until we really need to modify it, it doesn't need to be mutable

File: src/cli.rs [hunk 5]
```diff
@@ -102,96 +92,68 @@ pub fn init() {
...
+            "udpb:255.255.255.255:9999",
+            "udpc:127.0.0.1:1234",
+            "udpc:192.168.1.100:8080",
+            "udps:0.0.0.0:5000",
+            "udps:192.168.1.5:6789",
+        ];
+
+        for endpoint in endpoints {
+            dbg!(endpoints_parser(endpoint).is_ok());
+        }
```
Comment (joaoantoniocardoso) [inline src/cli.rs:157]: I'm assuming this is a development tool, not a test finished test. A suggestion for test: having a Vec of `(endpoint: &str, driver: impl DriverExt)` and asserting each `endpoints_parser(endpoint)` call results in the expected Ok(driver).

**Review by joaoantoniocardoso — COMMENTED** [review]
great work!

File: src/cli.rs (hunk 1 above)
Comment (patrickelectric) [reply src/cli.rs:64]: duh

File: src/cli.rs (hunk 5 above)
Comment (patrickelectric) [reply src/cli.rs]: Check now

**Review by joaoantoniocardoso — APPROVED** [review]

File: src/drivers/mod.rs [hunk 6]
```diff
@@ -17,3 +38,137 @@ pub trait Driver: Send + Sync {
...
+        ExtInfo {
+            driver_ext: Box::new(udp::client::UdpClientExt),
+            typ: Type::UdpClient,
+        },
+        ExtInfo {
+            driver_ext: Box::new(udp::server::UdpServerExt),
+            typ: Type::UdpServer,
+        },
+    ]
+}
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/drivers/mod.rs:174]: nice

File: src/drivers/mod.rs [hunk 7]
```diff
@@ -17,3 +38,137 @@ pub trait Driver: Send + Sync {
...
+fn process_old_format(entry: &str) -> Option<DriverDescriptionLegacy> {
+    let captures = Regex::new(r"^(?P<scheme>\w+):(?P<arg1>[^:]+)(:(?P<arg2>\d+))?$")
+        .unwrap()
+        .captures(entry)?;
+
+    let prefix = captures.name("scheme").unwrap().as_str();
+    let arg1 = captures.name("arg1").unwrap().as_str();
+    let arg2 = captures.name("arg2").map(|m| m.as_str());
+
+    let endpoints = endpoints(); // Assuming `endpoints` is a function returning Vec<ExtInfo>
```
[the remark this answers is not in the document]
Comment (patrickelectric) [reply src/drivers/mod.rs:102]: yep

**Pull request merged**